core/trino-main/src/main/java/io/trino/execution/scheduler/FullNodeCapableNodeAllocatorService.java

```diff
@@ -0,0 +1,595 @@
+                            fullNodePendingAcquires.remove(reservedNode);
+                            verify(fullNodesByQueryId.remove(pendingAcquire.getQueryId(), reservedNode));
+                            allocatedFullNodes.add(opportunisticNode.get());
+                            verify(fullNodesByQueryId.put(pendingAcquire.getQueryId(), opportunisticNode.get()));
+                            assignedNodes.put(pendingAcquire, opportunisticNode.get());
+                        }
+                        continue;
+                    }
+
+                    if (!currentCandidates.getCandidates().contains(reservedNode)) {
+                        // current candidate is gone; move pendingAcquire to detached state
+                        detachedFullNodePendingAcquires.add(pendingAcquire);
+                        fullNodePendingAcquires.remove(reservedNode);
+                        // trigger one more round of processing immediatelly
+                        wakeupProcessPendingAcquires();
+                    }
+
+                    // we are good acquiring reserved full node
+                    allocatedFullNodes.add(reservedNode);
+                    fullNodePendingAcquires.remove(reservedNode);
+                    assignedNodes.put(pendingAcquire, reservedNode);
+                }
+                catch (RuntimeException e) {
+                    failures.put(pendingAcquire, e);
+                    fullNodePendingAcquires.remove(reservedNode);
+                }
+            }
+        }
+
+        // complete futures outside of synchronized section
+        checkState(!Thread.holdsLock(this), "Cannot complete node futures under lock");
+        assignedNodes.forEach((pendingAcquire, node) -> {
+            SettableFuture<InternalNode> future = pendingAcquire.getFuture();
+            future.set(node);
+            if (future.isCancelled()) {
+                releaseFullNode(node, pendingAcquire.getQueryId());
+            }
+        });
+
+        failures.forEach((pendingAcquire, failure) -> {
+            SettableFuture<InternalNode> future = pendingAcquire.getFuture();
+            future.setException(failure);
+        });
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        stopped.set(true);
+        executor.shutdownNow();
+    }
+
+    public synchronized Optional<InternalNode> tryAcquire(NodeRequirements requirements, Candidates candidates, QueryId queryId)
+    {
+        if (isFullNode(requirements)) { // todo
+            return tryAcquireFullNode(candidates, queryId);
+        }
+        return tryAcquireSharedNode(candidates, requirements.getMemory().orElseThrow().toBytes());
+    }
+
+    private synchronized Optional<InternalNode> tryAcquireFullNode(Candidates candidates, QueryId queryId)
+    {
+        Collection<InternalNode> queryFullNodes = fullNodesByQueryId.get(queryId);
+
+        if (fullNodesCountExceeded(queryFullNodes.size(), candidates.getAllNodesCount())) {
```

Comment:
It the maximum allowed number of full nodes is reached for a certain query, but a query has more "full node" tasks to schedule should we try to reuse the nodes already assigned for a given query (to avoid waiting for a new node to free)?

Reply:
It should happen kinda "automatically". When we free the full node the `processPendingAcquires` routine will be woken up. It runs full node processing first (`processFullNodePendingAcquires`) so the pending allocation will be assigned to freed node via the loop going over `detachedFullNodePendingAcquires.iterator();`